### PR TITLE
feat(web-analytics): Add Sessions Table V2

### PR DESCRIPTION
**Pull request by robbie-c**
~~Don't merge this until https://github.com/PostHog/posthog/pull/22954 is in, as this PR uses that as a base~~

This PR currently has some failing tests, these tests are failing on master and are not introduced by this PR. I'll rebase when they are fixed. See https://posthog.slack.com/archives/C0113360FFV/p1718961268054819

Before this PR gets merged, we should change the where clause of the MV to only include events in the future. I'll do this as a ~last step before merging, as we might need to change it depending on when this gets merged.

## Problem

The previous session design was not fast enough, and did not allow for sampling. See https://github.com/PostHog/product-internal/pull/601

## Changes

* Add migration that adds the v2 raw_sessions table
* Add v2 raw_sessions and lazy sessions tables to hogql
* Add hogl modifier to enable this on a per-customer basis
* Update generate_demo_data to create events with a v7 sessions table
* Update tests

The biggest difference between the v1 and v2 tables is that v2 relies on the session_id being a uuid7, and can use this assumption to have a much better index. When just querying the session table alone for date range queries, this was around twice as fast, and should be even more of an upgrade when joining to the events table.

It also will support sampling, I expect we will have to tune the index granularity to work well but in theory we should be able to get very fast queries working, sampling by session.

## Does this work well for both Cloud and self-hosted?

Yes

## How did you test this code?

Added tests for the new mode, kept running tests against current mode

### CODE REVIEW

File: posthog/hogql/modifiers.py
Comment (robbie-c) [inline posthog/hogql/modifiers.py:56]: DON'T MERGE WITHOUT CHANGING THIS TO V1

Comment (robbie-c) [reply posthog/hogql/modifiers.py:56]: It's set to V2 just to ensure that tests can pass with this set to V2, though I'm also parametrizing the most relevant ones

File: posthog/models/raw_sessions/sql.py
Comment (robbie-c) [inline posthog/models/raw_sessions/sql.py:325]: Before we merge, add a date to this line

File: posthog/models/raw_sessions/sql.py
Comment (fuziontech) [inline posthog/models/raw_sessions/sql.py:13]: Heads up! This will work for now, but in the future something to know:

If there is any prod data in this table this will fail due to `max_table_size_to_drop` setting
https://clickhouse.com/docs/en/operations/server-configuration-parameters/settings#max_table_size_to_drop

We use this to protect from accidental deletions. Protects us from `DROP`s and `TRUNCATES`

File: posthog/models/raw_sessions/sql.py
Comment (fuziontech) [inline posthog/models/raw_sessions/sql.py:1]: I really like how this is organized. :+1: :+1:

**Review by fuziontech — APPROVED** [review]
Looks good! Added a comment, but other than that this looks solid. Excited to see it in action.

File: posthog/models/raw_sessions/sql.py
Comment (robbie-c) [reply posthog/models/raw_sessions/sql.py:13]: Good to know, the delete/truncates are only used in test setup code so should be ok

**Comment by Twixes** [comment]
I see Paul reviewed https://github.com/PostHog/product-internal/pull/601, so he probably has some context already – tagged him here.

File: posthog/hogql/test/test_parser_cpp.py
Comment (Twixes) [inline posthog/hogql/test/test_parser_cpp.py:8]: Oh interesting, I don't think this was needed before, so curious how you stumbled upon the need

**Review by Twixes — APPROVED** [review]
The code looks good to me, but sadly I don't have any context on sessions table v2

File: posthog/hogql/test/test_parser_cpp.py
Comment (robbie-c) [reply posthog/hogql/test/test_parser_cpp.py:8]: Nothing major, I just wanted to get the green arrow to appear in the UI, when I was trying to figure out when these tests were broken on main

File: posthog/models/raw_sessions/sql.py
Comment (robbie-c) [reply posthog/models/raw_sessions/sql.py]: Added a comment and changed the date

**Pull request merged**
